openedx/core/djangoapps/plugins/docs/decisions/0004-plugin-slots.rst

```diff
@@ -0,0 +1,103 @@
+Plugin Slots
+------------
+
+Status
+======
+Draft
+
+Context
+=======
+edx-platform contains a plugin system (https://github.com/edx/edx-platform/tree/master/openedx/core/djangoapps/plugins)
+which allows new Django apps to be installed inside the LMS and Studio without
+requiring the LMS/Studio to know about them. This is what enables us to move to
+a small and extensible core. While it's possible to extend the content of pages
+rendered by the platform using templates, via certain extension points that allow
+injecting content into 'head-extra', 'body-initial' etc slots in the base template,
+it isn't possible for plugins to inject content at all.
+
+Decisions
+=========
+We have added the ability for plugins to render content into existing pages. To
+support this, we have decided:
+
+* A template can how declare slots into which a plugin can inject content, by
+  using the `plugin_slot` template tag (for Django template) or function (for
+  Mako templates).
+* Plugins can define a callable function that the LMS or Studio can import and
+  call. This function will be called with minimal context, and in turn supports
+  pluggable contexts.
+* The callable function should return direct HTML content as text that can be
+  rendered on page.
+* Each view can whitelist what context data should be made available to all
+  plugins by adding it to the context itself in a list called `context_whitelist`.
+  This will need to be maintained across releases so should be kept to a bare
+  minimum.
+* All templates/pages will support three slots:
+   + ``head-extra``: This slot exists near the end of the header tag for each page
+     and can be used to add scripts, metadata, stylesheets or other header content.
+     It is equivalent to adding a 'head-extra.html' template file.
+   + ``body-initial``: This slot exists at the start of the page, right after the
+     opening of the body tag. It is equivalent to adding a 'body-initial.html'
+     template file.
+   + ``body-extra``: This slot exists at the end of the page near the closing of
+     the body tag. It is equivalent to adding a 'body-extra.html' template file.
+
+Implementation
+==============
+
+In the plugin app
+~~~~~~~~~~~~~~~~~
+
+Config
+++++++
+
+Inside of the AppConfig of your new plugin app, add a "slots_config" item like below.
+
+* The format will be ``{"slot_name": "function_inside_plugin_app"}``
+* The function name & path don't need to be named anything specific, so long as they work
+* These functions will be called on **every** render of that view, so keep them
```

Comment:
@ormsbee I'm curious what you think of this approach, in terms of sandboxing of misbehaving plugins.  Is there a safer way we could allow plugins to render HTML content to be injected into the page?  Or is the warning on efficiency/memoization here the best we can do?  

I'm thinking about your presentation about misbehaving ~puppies~ plugins.

Reply:
As is, this would make our response rendering performance directly beholden to the efficiency of our plugins.  I don't see a great way around it while still enabling this sort of behavior, though.